packages/vite/src/node/plugins/css.ts

```diff
@@ -2023,3 +2090,162 @@ const preProcessors = Object.freeze({
 function isPreProcessor(lang: any): lang is PreprocessLang {
   return lang && lang in preProcessors
 }
+
+const importLightningCSS = createCachedImport(() => import('lightningcss'))
+
+async function compileLightningCSS(
+  id: string,
+  src: string,
+  config: ResolvedConfig,
+  cssConfig: ResolvedLightningCSSOptions,
+  urlReplacer?: CssUrlReplacer,
+): ReturnType<typeof compileCSS> {
+  const deps = new Set<string>()
+  const res = styleAttrRE.test(id)
+    ? (await importLightningCSS()).transformStyleAttribute({
+        filename: id,
+        code: Buffer.from(src),
+        targets: cssConfig.targets,
+        minify: config.isProduction && config.build.cssMinify,
+        analyzeDependencies: true,
+      })
+    : await (
+        await importLightningCSS()
+      ).bundleAsync({
+        filename: id,
+        resolver: {
+          read(filename) {
+            if (filename === id) {
+              // TODO: add source map inline comment
+              return src
+            }
+            return fs.readFileSync(filename, 'utf-8')
+          },
+          async resolve(id, from) {
+            const publicFile = checkPublicFile(id, config)
+            if (publicFile) {
+              return publicFile
+            }
+
+            const resolved = await getAtImportResolvers(config).css(id, from)
+
+            if (resolved) {
+              deps.add(resolved)
+              return resolved
+            }
+            return id
+          },
+        },
+        targets: cssConfig.targets,
+        minify: config.isProduction && config.build.cssMinify,
+        sourceMap: true,
+        analyzeDependencies: true,
+        cssModules: cssModuleRE.test(id),
+        drafts: { nesting: true },
+      })
+
+  let css = res.code.toString()
+  for (const dep of res.dependencies!) {
+    switch (dep.type) {
+      case 'url':
+        deps.add(dep.url)
+        if (urlReplacer) {
+          css = css.replace(dep.placeholder, await urlReplacer(dep.url, id))
+        }
+        break
+      default:
+        throw new Error(`Unsupported dependency type: ${dep.type}`)
+    }
+  }
+  return {
+    code: css,
+    map: 'map' in res ? res.map?.toString() : undefined,
+    deps,
+    modules:
+      'exports' in res && res.exports
+        ? Object.fromEntries(
+            Object.entries(res.exports)
+              // https://github.com/parcel-bundler/lightningcss/issues/291
+              .sort((a, b) => a[0].localeCompare(b[0]))
+              .map(([key, value]) => [key, value.name]),
```

Comment:
This will need to handle CSS module [composition](https://lightningcss.dev/css-modules.html#class-composition). The `value.composes` property will contain a list of composed class names, so to get the final class name you need to join the strings by spaces. Since you are using the `bundle` API, dependencies between files are already taken care of, so the type of each composed class will be either "local" or "global".

Reply:
So should this be `[value.name, ...value.composes.map(c => c.name)].join(" ")` ? (I will rewrite this block to avoid so many array allocation)